presto-common/src/main/java/com/facebook/presto/common/block/VariableWidthBlockBuilder.java

```diff
@@ -0,0 +1,466 @@
+    {
+        this.blockBuilderStatus = blockBuilderStatus;
+
+        initialEntryCount = expectedEntries;
+        initialSliceOutputSize = min(expectedBytes, MAX_ARRAY_SIZE);
+
+        updateArraysDataSize();
+    }
+
+    @Override
+    public int getPositionOffset(int position)
+    {
+        checkValidPosition(position, positions);
+        return getOffset(position);
+    }
+
+    @Override
+    public int getSliceLength(int position)
+    {
+        checkValidPosition(position, positions);
+        return getOffset((position + 1)) - getOffset(position);
+    }
+
+    @Override
+    public Slice getRawSlice(int position)
+    {
+        return sliceOutput.getUnderlyingSlice();
+    }
+
+    @Override
+    public int getPositionCount()
+    {
+        return positions;
+    }
+
+    @Override
+    public long getSizeInBytes()
+    {
+        long arraysSizeInBytes = (Integer.BYTES + Byte.BYTES) * (long) positions;
+        return sliceOutput.size() + arraysSizeInBytes;
+    }
+
+    @Override
+    public long getRegionSizeInBytes(int positionOffset, int length)
+    {
+        int positionCount = getPositionCount();
+        checkValidRegion(positionCount, positionOffset, length);
+        long arraysSizeInBytes = (Integer.BYTES + Byte.BYTES) * (long) length;
+        return getOffset(positionOffset + length) - getOffset(positionOffset) + arraysSizeInBytes;
+    }
+
+    @Override
+    public long getPositionsSizeInBytes(boolean[] positions)
+    {
+        checkValidPositions(positions, getPositionCount());
+        long sizeInBytes = 0;
+        int usedPositionCount = 0;
+        for (int i = 0; i < positions.length; ++i) {
+            if (positions[i]) {
+                usedPositionCount++;
+                sizeInBytes += getOffset(i + 1) - getOffset(i);
+            }
+        }
+        return sizeInBytes + (Integer.BYTES + Byte.BYTES) * (long) usedPositionCount;
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
+    {
+        long size = INSTANCE_SIZE + sliceOutput.getRetainedSize() + arraysRetainedSizeInBytes;
+        if (blockBuilderStatus != null) {
+            size += BlockBuilderStatus.INSTANCE_SIZE;
+        }
+        return size;
+    }
+
+    @Override
+    public void retainedBytesForEachPart(BiConsumer<Object, Long> consumer)
+    {
+        consumer.accept(sliceOutput, sliceOutput.getRetainedSize());
+        consumer.accept(offsets, sizeOf(offsets));
+        consumer.accept(valueIsNull, sizeOf(valueIsNull));
+        consumer.accept(this, (long) INSTANCE_SIZE);
+    }
+
+    @Override
+    public Block copyPositions(int[] positions, int offset, int length)
+    {
+        checkArrayRange(positions, offset, length);
+
+        int finalLength = 0;
+        for (int i = offset; i < offset + length; i++) {
+            finalLength += getSliceLength(positions[i]);
+        }
+        SliceOutput newSlice = Slices.allocate(finalLength).getOutput();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *newSlice*. Execution paths that do not contain closure statements were detected.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).